_posts/2024-08-27-building-browser.md

```diff
@@ -0,0 +1,174 @@
+We will also define more specific webview types to satisfy different purposes in the future.
+
+## Multi-window
+
+The other prominent feature after multi-view is the ability to support multiple windows.
+This one wasn’t planned at first, but because it affects too many components, we ended up resolving them together from the ground up.
+
+Servo uses [WebRender](https://github.com/servo/webrender), based on OpenGL, to render its layout.
+To support multiple windows, we need to support multiple OpenGL surfaces.
+One approach would be to create separate OpenGL contexts for each window.
+But since our implementations of WebGL, WebGPU, and WebXR are all tied to a single WebRender instance, which in turn only supports a single OpenGL context for now, we chose to use a single context with multiple surfaces.
+This alternative approach could potentially use less memory and spawn fewer threads.
+For more details, see [this series of blog posts](https://wusyong.github.io/posts/verso-compositor-part3/) by [@wusyong](https://github.com/wusyong).
+
+<figure class="_fig"><a href="{{ '/img/blog/multiwindow.png' | url }}"><img src="{{ '/img/blog/multiwindow.png' | url }}"
+    alt="Verso displaying two windows"></a>
+<figcaption>Figure 2: Verso creates two separate windows with the same OpenGL context.</figcaption></figure>
+
+There is still room for improvement.
+For example, WebRender currently [only supports rendering a single “document”](https://github.com/servo/webrender/blob/0.64/webrender/src/renderer/mod.rs#L1228).
+Unless we create multiple WebRender instances, like Firefox does, we have one WebRender document that has to constantly update all of its display lists to show on all of our windows.
+This could potentially lead to race conditions where a webview may draw to the wrong window for a split second.
+
+There are also different OpenGL versions across multiple platforms, which can be challenging to configure and link.
+Verso is experimenting with using [Glutin](https://docs.rs/glutin/latest/glutin/) for better configuration and attempting to get closer to the general Rust ecosystem.
+
+## What’s next?
+
+With multi-view and multi-window support as the fundamental building blocks, we could create more UI elements to keep pushing the envelope of our browser and embedding research.
+At the same time, Servo is a huge project, with many potential improvements still to come, so we want to reflect on our progress and decide on our priorities.
+Here are some directions that are worth pursuing.
+
+### Are we browser yet?
+
+We are building a page, [arewebrowseryet.com](https://servo.zulipchat.com/#narrow/stream/263398-general/topic/arewebrowseryet.2Ecom), in the hope of gathering the strength of the community to help us display the statistics of supported CSS properties and web APIs in Servo by [popularity order](https://chromestatus.com/metrics/css/popularity) and benchmark results such as jetstream2 and speedometer3.
```

Comment:
It can be nice to mention this idea, but I'm afraid if we could be over promising here. Maybe we could wait until we have something more detailed to announce this. Also this is not strictly related to this blog post, so could go in a monthly update when the website is ready.

Also the link to Zulip is not public if you don't have a Zulip account, so it's not usually nice to find links that go to a page where you need to register.

Reply:
I update the paragraph a bit in Added in [f0e7fa9](https://github.com/servo/servo.org/pull/171/commits/f0e7fa9024bb157c7bfd66bda46bd04d9a84522e).
It focuses more on benchmarking and statistic metrics now. I also added what sagudev has done so far.

Reply:
Thanks, this looks better to me.